### ISSUE TITLE
Network throttling in iOS app programatically for automation.

### ISSUE THREAD

**Issue by amit1990sumit** [h2non/toxy#59]
Hi Team,

Is it possible that I can throttle network on my iOS app or specific URL?
Please help me to understand this, what are the things required for this if possible.

Thanks

**Comment by h2non** [comment]
Yes, as long as you put `toxy` in the middle of the communication:
```
iOS --> toxy proxy --> target server
```

Your iOS should communicate with the `toxy` proxy directly and there is where the throttling would happen.

**Comment by amit1990sumit** [comment]
Thanks for reply. Since I am new to node js, could you please guide me could hit api to localhost using node js?

Thanks again

**Comment by h2non** [comment]
See some examples, such as: https://github.com/h2non/toxy/blob/master/examples/bandwidth-limit.js

Run it locally from the console via: `$ node example.js`

You need to replace the `httpbin.org` domain with the server you want to consume. Afterwards, you need to consume the toxy server running in localhost in your iOS app. Assuming you can't consume it via local IP address/hostname in your LAN, you can use this node.js package to create a tunnel in order to expose the toxy server to the Internet, and consume from everywhere, including your iOS app:
```bash
$ npm install -g localtunnel
```
Then:
```bash
$ lt --port 3000 --subdomain my-toxy-proxy-server
```

That's all. Enjoy it.

**Comment by amit1990sumit** [comment]
Thank you so much for ur reply.
One last question, can I add proxy ip and port inside iphone proxy setting so that we can track all the network call which go through iphone. I want same as charles proxy, only thing it should be programmatically.
Please help me.

**Comment by amit1990sumit** [comment]
Thank you so much for ur reply.
One last question, can I add proxy ip and port inside iphone proxy setting so that we can track all the network call which go through iphone. I want same as charles proxy, only thing it should be programmatically.
Please help me.

**Comment by h2non** [comment]
As far as I know, you don't need to explicitly use the "proxy" mode in the iPhone. You just need to communicate to the target IP/hostname of `toxy`, and then you can programmatically inspect, log, modify the incoming HTTP traffic from the phone. You can see the middleware/intercept/logging examples.

**Comment by amit1990sumit** [comment]
Ok, but i do not want to proxy specific url, In my case, my framework itself hit all the network call and i am unable to access the urls. I want to add proxy so that any network call able to track which go through my iphone.

Please help and thank you again

**Comment by amit1990sumit** [comment]
Today, I tried all the steps what you mentioned, 
1) I create the tunnel and run "lt --port 3000 --subdomain my-toxy-proxy-server" which gave URL = "https://my-toxy-proxy-server.localtunnel.me"
2) Replaced the same URL in example project "bandwidth-limit.js".
3) run the "bandwidth-limit.js" which gives "Server listening on port: 3000". 

but found that no call goes through a proxy. 

Currently, I am trying this in a browser. Please let me know if anything missing from my end. Please help me.

**Comment by h2non** [comment]
This looks like an XY problem. I'm not actually sure that you know what you want or you actually understand how the network works. Please, make sure you understand these things. Otherwise I can't help.

**Comment by amit1990sumit** [comment]
Thanks for reply and sorry to bother you.
Actually I am new to server side script and most of the thing I undertood now.
Only prob I am facing now, In readme file there are serveral API mentioned.

```
API
Hierarchy:

Servers - Managed toxy instances
Rules - Globally applied rules
Poisons - Globally applied poisons
Rules - Poison-specific rules
Routes - List of configured routes
Route - Object for each specific route
Rules - Route-level registered rules
Poisons - Route-level registered poisons
Rules - Route-level poison-specific rules
GET /
Servers
GET /servers
GET /servers/:id
Rules
GET /servers/:id/rules
POST /servers/:id/rules
Accepts: application/json

Example payload:

{
  "name": "method",
  "options": "GET"
}
```

Just want to know how I can access these API from client end.

**Comment by h2non** [comment]
You can take a look at the instructions on the Readme > Admin API section. Basically, you need to start 2 HTTP servers running on different ports, one for the proxy itself, and another for the HTTP Admin API. You can see an example as well on the repository. Note that this is apposite to your initial intention about having programmatic control of the proxy.

**Comment by amit1990sumit** [comment]
Thanks for the reply.
I am able to run two server and able to pass the request through a proxy server. Also, at runtime programmatically I am able to hit the HTTP API request to set Poisons and Rules according to my requirement and it is working for me like to generate an error in between.

Now the only issue, when I hit HTTP API to set limited bandwidth, it is not setting the actual bandwidth what I am passing to in json request mentioned below.
Using: POST  http://localhost:9000/servers/8dd/poisons (I am using the simulator for now)
{
  "name": "bandwidth",
  "phase": "outgoing",
  "option": { "bps": 512000}
}
Thanks again for helping me throughout.
